src/main/java/fact/io/hdureader/BinTable.java

```diff
@@ -0,0 +1,256 @@
+                .filter(entry -> entry.getKey().matches("TTYPE\\d+"))
+                .sorted(Comparator.comparing(e -> Integer.parseInt(e.getKey().substring(5))))
+                .collect(Collectors.toList());
+
+
+        ttypes.forEach(entry -> {
+            final int n = Integer.parseInt(entry.getKey().substring(5));
+            HeaderLine zform = entries.stream()
+                    .filter(e -> e.getKey().matches("ZFORM" + n))
+                    .findFirst()
+                    .map(Map.Entry::getValue)
+                    .orElse(null);
+
+            HeaderLine tform = entries.stream()
+                    .filter(e -> e.getKey().matches("TFORM" + n))
+                    .findFirst()
+                    .map(Map.Entry::getValue)
+                    .orElse(null);
+
+
+            columns.add(new TableColumn(zform, tform, entry.getValue()));
+        });
+
+
+        numberOfRowsInTable = header.getInt("NAXIS2").orElse(0);
+        numberOfColumnsInTable = columns.size();
+
+
+        // create readers for the data in this table
+        //create stream for data in table area of the bintable extension
+        DataInputStream stream = Fits.getDecompressedDataStream(url);
+
+        long skippedBytes = stream.skipBytes((int) (hduOffset + header.headerSizeInBytes));
+        if(skippedBytes != hduOffset + header.headerSizeInBytes){
+            throw new IOException("Could not skip all bytes to table data in this HDU.");
+        }
+        this.tableDataStream = stream;
+
+
+        //check if this bintable has data following after the table. See section 7.3.1 in the fits standard
+        long pcount = header.getLong("PCOUNT").orElse(0L);
+
+        if (pcount > 0) {
+            Integer naxis1 = header.getInt("NAXIS1").orElse(0);
+            Integer naxis2 = header.getInt("NAXIS2").orElse(0);
+
+            int theap = header.getInt("ZHEAPPTR").orElseGet(() -> header.getInt("THEAP").orElse(naxis1*naxis2));
+
+            int bytesToSkip = Math.toIntExact((hduOffset + header.headerSizeInBytes + theap));
+
+            DataInputStream hStream = Fits.getDecompressedDataStream(url);
+            skippedBytes = hStream.skipBytes(bytesToSkip);
+
+            if (skippedBytes != bytesToSkip) {
+                throw new IOException("Could not skip all bytes to heap area of the bintable in this HDU.");
+            }
+            this.heapDataStream = hStream;
+        }
+    }
+
+
+    /**
+     * Perform some sanity checks on the bintable.
+     *  1. The TFIELDS keyword must exist.
+     *  2. The number of columns stored in the TFIELD line has to fit the number
+     *     of TTPYE fields
+     *  3. Make sure the NAXIS2 keyword exists. It gives the number of rows.
+     *  4. NAXIS2 needs to be non-negative.
+     *
+     * @param header the header to check
+     */
+    private void binTableSanityCheck(Header header) {
+        int tfields = header.getInt("TFIELDS").orElseThrow(() -> {
```

Comment:
`tfields` here is equal to (and has to be equal to) `numberOfColumnsInTable = columns.size();`
further above. I have the feeling some stuff is done in a duplicate way here. This might be a little confusing to the reader...